# Telecomm EDA project

### Import Libraries

In [2]:
!pip install folium
!pip install geopy

In [38]:
import numpy as np
import pandas as pd
import folium as folium
from geopy.geocoders import Nominatim

%matplotlib inline

# Cleanup

### Turn the CSV into a PD Dataframe

In [4]:
df = pd.read_csv("datasets/metadata.csv").drop(["Cell Cgi"], axis = 1)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10476 entries, 0 to 10475
Data columns (total 6 columns):
Cell Tower Location     10476 non-null object
Comm Identifier         1374 non-null object
Comm Timedate String    10476 non-null object
Comm Type               10476 non-null object
Latitude                10476 non-null float64
Longitude               10476 non-null float64
dtypes: float64(2), object(4)
memory usage: 491.1+ KB


### Turn Datetime String into a numpy Datetime Object

In [5]:
df["Datetime"] = pd.to_datetime(df["Comm Timedate String"], infer_datetime_format=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10476 entries, 0 to 10475
Data columns (total 7 columns):
Cell Tower Location     10476 non-null object
Comm Identifier         1374 non-null object
Comm Timedate String    10476 non-null object
Comm Type               10476 non-null object
Latitude                10476 non-null float64
Longitude               10476 non-null float64
Datetime                10476 non-null datetime64[ns]
dtypes: datetime64[ns](1), float64(2), object(4)
memory usage: 573.0+ KB


### Make day of week column

In [6]:
df['Day of Week'] = df['Datetime'].dt.weekday_name

### Make the coordinates Folium Friendly :)

In [7]:
df['Location']=list(zip(df.Latitude,df.Longitude))

### Cleanup and reorganize DataFrame

In [9]:
df = df[['Datetime', 'Day of Week', 'Comm Identifier', 'Comm Type', 'Cell Tower Location', 'Location']]
df.head()

Datetime Day of Week                           Comm Identifier  \
0 2014-04-01 09:40:00     Tuesday  f1a6836c0b7a3415a19a90fdd6f0ae18484d6d1e   
1 2014-04-01 09:42:00     Tuesday  62157ccf2910019ffd915b11fa037243b75c1624   
2 2014-04-01 13:13:00     Tuesday  c8f92bd0f4e6fb45ed7fce96fc831b283db2b642   
3 2014-04-01 13:13:00     Tuesday  f1a6836c0b7a3415a19a90fdd6f0ae18484d6d1e   
4 2014-04-01 17:27:00     Tuesday  f1a6836c0b7a3415a19a90fdd6f0ae18484d6d1e   

  Comm Type Cell Tower Location                     Location  
0     Phone          REDFERN TE  (-33.89293336, 151.2022962)  
1     Phone          REDFERN TE  (-33.89293336, 151.2022962)  
2     Phone         HAYMARKET #  (-33.88032891, 151.2056904)  
3     Phone         HAYMARKET #  (-33.88032891, 151.2056904)  
4     Phone         HAYMARKET #  (-33.88032891, 151.2056904)

### Shorten Comm Identifier

In [10]:
df["Comm Identifier"] = df["Comm Identifier"].str[:10]

df.head()

Datetime Day of Week Comm Identifier Comm Type  \
0 2014-04-01 09:40:00     Tuesday      f1a6836c0b     Phone   
1 2014-04-01 09:42:00     Tuesday      62157ccf29     Phone   
2 2014-04-01 13:13:00     Tuesday      c8f92bd0f4     Phone   
3 2014-04-01 13:13:00     Tuesday      f1a6836c0b     Phone   
4 2014-04-01 17:27:00     Tuesday      f1a6836c0b     Phone   

  Cell Tower Location                     Location  
0          REDFERN TE  (-33.89293336, 151.2022962)  
1          REDFERN TE  (-33.89293336, 151.2022962)  
2         HAYMARKET #  (-33.88032891, 151.2056904)  
3         HAYMARKET #  (-33.88032891, 151.2056904)  
4         HAYMARKET #  (-33.88032891, 151.2056904)

# EDA

### Where this individual spends the most time.

They can usually be found in Sydney Australia.

- spends both weekend and weekday time near Amazonia Martial Arts, his top location overall.  this is likely his home
- spends a lot of weekday time at IFGF Sydney (a church), maybe volunteering?
- parks at City Parking in Haymarket on weekdays, rarely on weekends.  this is probably near where he works

In [11]:
df['Location'].dropna().value_counts().head(10)

(-33.78815, 151.26654)         4301
(-33.88417103, 151.20235)      1084
(-42.84338, 147.29569)          723
(-33.89293336, 151.2022962)     712
(-33.88032891, 151.2056904)     563
(-42.85984, 147.29215)          501
(-33.779333, 151.276901)        465
(-33.79661, 151.27756)          454
(-33.796679, 151.285293)        231
(-42.85307, 147.31532)          197
Name: Location, dtype: int64

### 4301
-33.788150, 151.266540 = Amazonia Martial Arts Club (Manly Beach 

### 1084
-33.884171, 151.202350 = IFGF Sydney (a church) (Haymarket, Sydney)

### 723
-42.843380, 147.295690 = Shopping Mall, Intercity Cycleway (Tasmania)

### 712
-33.892933, 151.202296 = Office Building outside fire station (Redfern, Sydney)

### 563
-33.880329, 151.205690 = City Parking (Haymarket, Sydney)

There are four distinct periods of time where the individual goes to Tasmania: 
- 5/2/2014 to 5/4/2014
- 8/10/2014 to 8/12/2014
- 11/12/2014 to 11/15/2014
- 12/20/2014 to 12/26/2014

### Weekdays vs Weekends

In [12]:
# build a mask to select days 5 and 6 (saturday, sunday)
df.loc[~df.Datetime.dt.weekday.isin([5,6])]

Datetime Day of Week Comm Identifier Comm Type  \
0     2014-04-01 09:40:00     Tuesday      f1a6836c0b     Phone   
1     2014-04-01 09:42:00     Tuesday      62157ccf29     Phone   
2     2014-04-01 13:13:00     Tuesday      c8f92bd0f4     Phone   
3     2014-04-01 13:13:00     Tuesday      f1a6836c0b     Phone   
4     2014-04-01 17:27:00     Tuesday      f1a6836c0b     Phone   
5     2014-04-01 17:36:00     Tuesday      6bbc17070a     Phone   
6     2014-04-01 17:40:00     Tuesday      6bbc17070a     Phone   
7     2014-04-02 19:18:00   Wednesday      7cb96eadd3     Phone   
8     2014-04-03 14:35:00    Thursday      de40c5c1f9     Phone   
9     2014-04-03 14:36:00    Thursday      66f32c1163     Phone   
10    2014-04-04 09:12:00      Friday      bc0b018604       SMS   
11    2014-04-04 09:39:00      Friday      62157ccf29     Phone   
12    2014-04-04 09:39:00      Friday      91aba4a113     Phone   
13    2014-04-04 09:47:00      Friday      91aba4a113     Phone   
14    2014-04-04 14:00:00      Friday      de40c5c1f9     Phone   
15    2014-04-04 15:11:00      Friday      de40c5c1f9     Phone   
16    2014-04-04 15:20:00      Friday      b5238f237b     Phone   
17    2014-04-04 18:10:00      Friday      91aba4a113     Phone   
18    2014-04-04 18:28:00      Friday      91aba4a113     Phone   
19    2014-04-04 18:31:00      Friday      91aba4a113     Phone   
20    2014-04-04 18:57:00      Friday      c8f92bd0f4     Phone   
21    2014-04-04 18:58:00      Friday      91aba4a113     Phone   
23    2014-04-07 10:46:00      Monday      62157ccf29     Phone   
24    2014-04-07 12:31:00      Monday      bc0b018604       SMS   
25    2014-04-07 18:18:00      Monday      70e1f163d8       SMS   
26    2014-04-07 18:23:00      Monday      12e3d1b0c9     Phone   
27    2014-04-07 18:31:00      Monday      6bbc17070a     Phone   
28    2014-04-07 18:39:00      Monday      70e1f163d8       SMS   
29    2014-04-07 18:39:00      Monday      70e1f163d8       SMS   
30    2014-04-07 18:54:00      Monday      70e1f163d8       SMS   
...                   ...         ...             ...       ...   
10446 2015-04-01 19:00:00   Wednesday      c13b4711eb     Phone   
10447 2015-04-01 19:07:00   Wednesday             NaN  Internet   
10448 2015-04-01 19:09:00   Wednesday             NaN  Internet   
10449 2015-04-01 19:09:00   Wednesday             NaN  Internet   
10450 2015-04-01 19:09:00   Wednesday             NaN  Internet   
10451 2015-04-01 19:20:00   Wednesday             NaN  Internet   
10452 2015-04-01 19:20:00   Wednesday             NaN  Internet   
10453 2015-04-01 19:25:00   Wednesday             NaN  Internet   
10454 2015-04-01 19:58:00   Wednesday             NaN  Internet   
10455 2015-04-01 20:13:00   Wednesday             NaN  Internet   
10456 2015-04-01 20:13:00   Wednesday             NaN  Internet   
10457 2015-04-01 20:15:00   Wednesday             NaN  Internet   
10458 2015-04-01 20:15:00   Wednesday             NaN  Internet   
10459 2015-04-01 20:16:00   Wednesday             NaN  Internet   
10460 2015-04-01 20:16:00   Wednesday      91aba4a113     Phone   
10461 2015-04-01 20:49:00   Wednesday             NaN  Internet   
10462 2015-04-01 20:49:00   Wednesday             NaN  Internet   
10463 2015-04-01 20:50:00   Wednesday             NaN  Internet   
10464 2015-04-01 21:03:00   Wednesday             NaN  Internet   
10465 2015-04-01 21:03:00   Wednesday             NaN  Internet   
10466 2015-04-01 21:03:00   Wednesday      91aba4a113     Phone   
10467 2015-04-01 21:04:00   Wednesday             NaN  Internet   
10468 2015-04-01 21:20:00   Wednesday             NaN  Internet   
10469 2015-04-01 21:20:00   Wednesday             NaN  Internet   
10470 2015-04-01 21:32:00   Wednesday             NaN  Internet   
10471 2015-04-01 22:40:00   Wednesday             NaN  Internet   
10472 2015-04-01 22:41:00   Wednesday             NaN  Internet   
10473 2015-04-01 22:44:00   Wednesday             NaN  Internet 

In [13]:
# build a mask to select weekdays
df.loc[df.Datetime.dt.weekday.isin([5,6])]

Datetime Day of Week Comm Identifier Comm Type  \
22    2014-04-05 15:10:00    Saturday      dc6774d10e     Phone   
50    2014-04-12 08:16:00    Saturday      0767e517fa     Phone   
51    2014-04-13 01:42:00      Sunday      6ce4ad58c4     Phone   
52    2014-04-13 10:58:00      Sunday      6bbc17070a     Phone   
53    2014-04-13 11:20:00      Sunday      cad22015a5       SMS   
54    2014-04-13 17:44:00      Sunday      6bbc17070a     Phone   
91    2014-04-19 12:34:00    Saturday      a5834ee77b     Phone   
135   2014-04-26 11:11:00    Saturday      a804558e42     Phone   
136   2014-04-26 11:47:00    Saturday      cd3b394668       SMS   
137   2014-04-26 11:49:00    Saturday      cd3b394668       SMS   
138   2014-04-26 11:55:00    Saturday      cd3b394668       SMS   
139   2014-04-26 11:55:00    Saturday      cd3b394668       SMS   
140   2014-04-26 11:57:00    Saturday      cd3b394668       SMS   
141   2014-04-26 11:58:00    Saturday      cd3b394668       SMS   
142   2014-04-26 12:14:00    Saturday      cd3b394668       SMS   
143   2014-04-26 12:21:00    Saturday      cd3b394668       SMS   
144   2014-04-26 12:34:00    Saturday      cd3b394668       SMS   
145   2014-04-26 12:38:00    Saturday      cd3b394668       SMS   
146   2014-04-27 11:35:00      Sunday      70e1f163d8     Phone   
147   2014-04-27 11:45:00      Sunday      70e1f163d8     Phone   
174   2014-05-04 10:46:00      Sunday      cd3b394668       SMS   
175   2014-05-04 16:12:00      Sunday      7b79c2b0f7     Phone   
176   2014-05-04 20:05:00      Sunday      91aba4a113     Phone   
177   2014-05-04 20:08:00      Sunday      91aba4a113     Phone   
178   2014-05-04 20:10:00      Sunday      12e3d1b0c9     Phone   
179   2014-05-04 20:10:00      Sunday      91aba4a113     Phone   
180   2014-05-04 20:43:00      Sunday      91aba4a113     Phone   
181   2014-05-04 20:51:00      Sunday      12e3d1b0c9     Phone   
217   2014-05-10 10:41:00    Saturday      bc0b018604       SMS   
218   2014-05-10 13:17:00    Saturday      dc6774d10e     Phone   
...                   ...         ...             ...       ...   
10169 2015-03-29 19:04:00      Sunday             NaN  Internet   
10170 2015-03-29 19:07:00      Sunday             NaN  Internet   
10171 2015-03-29 19:14:00      Sunday             NaN  Internet   
10172 2015-03-29 19:14:00      Sunday             NaN  Internet   
10173 2015-03-29 19:14:00      Sunday             NaN  Internet   
10174 2015-03-29 19:17:00      Sunday             NaN  Internet   
10175 2015-03-29 19:18:00      Sunday             NaN  Internet   
10176 2015-03-29 19:18:00      Sunday             NaN  Internet   
10177 2015-03-29 19:19:00      Sunday             NaN  Internet   
10178 2015-03-29 19:55:00      Sunday             NaN  Internet   
10179 2015-03-29 20:02:00      Sunday             NaN  Internet   
10180 2015-03-29 20:02:00      Sunday             NaN  Internet   
10181 2015-03-29 20:03:00      Sunday             NaN  Internet   
10182 2015-03-29 20:05:00      Sunday             NaN  Internet   
10183 2015-03-29 20:06:00      Sunday             NaN  Internet   
10184 2015-03-29 20:06:00      Sunday             NaN  Internet   
10185 2015-03-29 20:06:00      Sunday             NaN  Internet   
10186 2015-03-29 20:06:00      Sunday             NaN  Internet   
10187 2015-03-29 20:14:00      Sunday             NaN  Internet   
10188 2015-03-29 20:18:00      Sunday             NaN  Internet   
10189 2015-03-29 20:18:00      Sunday             NaN  Internet   
10190 2015-03-29 20:20:00      Sunday             NaN  Internet   
10191 2015-03-29 20:20:00      Sunday             NaN  Internet   
10192 2015-03-29 20:23:00      Sunday             NaN  Internet   
10193 2015-03-29 20:26:00      Sunday             NaN  Internet   
10194 2015-03-29 21:00:00      Sunday             NaN  Internet   
10195 2015-03-29 21:01:00      Sunday             NaN  Internet   
10196 2015-03-29 21:05:00      Sunday             NaN  Internet 

### Select comm type

In [16]:
# put which types you want to select in this list and select from it below
comm_types = ['Phone','SMS','Internet']

df[df['Comm Type'].isin(comm_types)]

Datetime Day of Week Comm Identifier Comm Type  \
0     2014-04-01 09:40:00     Tuesday      f1a6836c0b     Phone   
1     2014-04-01 09:42:00     Tuesday      62157ccf29     Phone   
2     2014-04-01 13:13:00     Tuesday      c8f92bd0f4     Phone   
3     2014-04-01 13:13:00     Tuesday      f1a6836c0b     Phone   
4     2014-04-01 17:27:00     Tuesday      f1a6836c0b     Phone   
5     2014-04-01 17:36:00     Tuesday      6bbc17070a     Phone   
6     2014-04-01 17:40:00     Tuesday      6bbc17070a     Phone   
7     2014-04-02 19:18:00   Wednesday      7cb96eadd3     Phone   
8     2014-04-03 14:35:00    Thursday      de40c5c1f9     Phone   
9     2014-04-03 14:36:00    Thursday      66f32c1163     Phone   
10    2014-04-04 09:12:00      Friday      bc0b018604       SMS   
11    2014-04-04 09:39:00      Friday      62157ccf29     Phone   
12    2014-04-04 09:39:00      Friday      91aba4a113     Phone   
13    2014-04-04 09:47:00      Friday      91aba4a113     Phone   
14    2014-04-04 14:00:00      Friday      de40c5c1f9     Phone   
15    2014-04-04 15:11:00      Friday      de40c5c1f9     Phone   
16    2014-04-04 15:20:00      Friday      b5238f237b     Phone   
17    2014-04-04 18:10:00      Friday      91aba4a113     Phone   
18    2014-04-04 18:28:00      Friday      91aba4a113     Phone   
19    2014-04-04 18:31:00      Friday      91aba4a113     Phone   
20    2014-04-04 18:57:00      Friday      c8f92bd0f4     Phone   
21    2014-04-04 18:58:00      Friday      91aba4a113     Phone   
22    2014-04-05 15:10:00    Saturday      dc6774d10e     Phone   
23    2014-04-07 10:46:00      Monday      62157ccf29     Phone   
24    2014-04-07 12:31:00      Monday      bc0b018604       SMS   
25    2014-04-07 18:18:00      Monday      70e1f163d8       SMS   
26    2014-04-07 18:23:00      Monday      12e3d1b0c9     Phone   
27    2014-04-07 18:31:00      Monday      6bbc17070a     Phone   
28    2014-04-07 18:39:00      Monday      70e1f163d8       SMS   
29    2014-04-07 18:39:00      Monday      70e1f163d8       SMS   
...                   ...         ...             ...       ...   
10446 2015-04-01 19:00:00   Wednesday      c13b4711eb     Phone   
10447 2015-04-01 19:07:00   Wednesday             NaN  Internet   
10448 2015-04-01 19:09:00   Wednesday             NaN  Internet   
10449 2015-04-01 19:09:00   Wednesday             NaN  Internet   
10450 2015-04-01 19:09:00   Wednesday             NaN  Internet   
10451 2015-04-01 19:20:00   Wednesday             NaN  Internet   
10452 2015-04-01 19:20:00   Wednesday             NaN  Internet   
10453 2015-04-01 19:25:00   Wednesday             NaN  Internet   
10454 2015-04-01 19:58:00   Wednesday             NaN  Internet   
10455 2015-04-01 20:13:00   Wednesday             NaN  Internet   
10456 2015-04-01 20:13:00   Wednesday             NaN  Internet   
10457 2015-04-01 20:15:00   Wednesday             NaN  Internet   
10458 2015-04-01 20:15:00   Wednesday             NaN  Internet   
10459 2015-04-01 20:16:00   Wednesday             NaN  Internet   
10460 2015-04-01 20:16:00   Wednesday      91aba4a113     Phone   
10461 2015-04-01 20:49:00   Wednesday             NaN  Internet   
10462 2015-04-01 20:49:00   Wednesday             NaN  Internet   
10463 2015-04-01 20:50:00   Wednesday             NaN  Internet   
10464 2015-04-01 21:03:00   Wednesday             NaN  Internet   
10465 2015-04-01 21:03:00   Wednesday             NaN  Internet   
10466 2015-04-01 21:03:00   Wednesday      91aba4a113     Phone   
10467 2015-04-01 21:04:00   Wednesday             NaN  Internet   
10468 2015-04-01 21:20:00   Wednesday             NaN  Internet   
10469 2015-04-01 21:20:00   Wednesday             NaN  Internet   
10470 2015-04-01 21:32:00   Wednesday             NaN  Internet   
10471 2015-04-01 22:40:00   Wednesday             NaN  Internet   
10472 2015-04-01 22:41:00   Wednesday             NaN  Internet   
10473 2015-04-01 22:44:00   Wednesday             NaN  Internet 

### Look at tower locations and build addresses

In [17]:
# make a dictionary to store addresses
address={}

# look up addresses and store them in the dictionary
geolocator = Nominatim()
for i in df.Location.unique() :
    location= geolocator.reverse(i)
    address[i]=location.address

In [18]:
# clean up the address
addressl={}
for k,v in address.items():
    fields=list(v.split(','))
    addressl[k]=fields

In [19]:
addressl

{(-42.884809999999995, 147.32748000000001): ["Domino's",
  ' Macquarie Street',
  ' Hobart',
  ' Tasmania',
  ' 7000',
  ' Australia'],
 (-42.88194, 147.32893000000001): ['Dôme',
  ' 29',
  ' Elizabeth Street',
  ' Hobart',
  ' Tasmania',
  ' 7000',
  ' Australia'],
 (-42.881009999999996, 147.33346): ['Hotel Grand Chancellor Hobart',
  ' 1',
  ' Davey Street',
  ' Hobart',
  ' Tasmania',
  ' 7000',
  ' Australia'],
 (-42.880290000000002, 147.32808): ['Dumpling World',
  ' Liverpool Street',
  ' Hobart',
  ' Tasmania',
  ' 7000',
  ' Australia'],
 (-42.874569999999999, 147.31788): ['Burnett St',
  ' Burnett Street',
  ' North Hobart',
  ' Tasmania',
  ' 7000',
  ' Australia'],
 (-42.860599999999998, 147.45419999999999): ['Mount Rumney Road',
  ' Acton Park',
  ' Mount Rumney',
  ' Tasmania',
  ' 7019',
  ' Australia'],
 (-42.859840000000005, 147.29214999999999): ['William Cooper Drive',
  ' New Town',
  ' Tasmania',
  ' 7008',
  ' Australia'],
 (-42.853070000000002, 147.31531999999999):

In [20]:
# make a field for each part of the address to sort and show

# df['address']=df.Location.map(lambda x: address[x])
df['Place']=df.Location.map(lambda x: addressl[x][0])
df['Town']=df.Location.map(lambda x: addressl[x][-4])
df['Region']=df.Location.map(lambda x: addressl[x][-3])
df['Zip']=df.Location.map(lambda x: addressl[x][-2])
df['Country']=df.Location.map(lambda x: addressl[x][-1])

In [22]:
# what unique regions has he/she visited?

df['Region'].unique()

array([' NSW', ' Tasmania', ' Victoria', ' Hume'], dtype=object)

### Select hour groupings (night, work, after work)

In [73]:
# from midnight to 8am
df[df.Datetime.dt.hour < 8]

# select top locations for this time period
df[df.Datetime.dt.hour < 8]['Location'].value_counts().head()

(-33.78815, 151.26654)         743
(-33.88417103, 151.20235)      128
(-33.79661, 151.27756)          48
(-33.779333, 151.276901)        45
(-33.88032891, 151.2056904)     40
Name: Location, dtype: int64

In [74]:
# from 8am-5pm
df[(df.Datetime.dt.hour > 7) & (df.Datetime.dt.hour < 17)]

# select top locations for this time period
df[(df.Datetime.dt.hour > 7) & (df.Datetime.dt.hour < 17)]['Location'].value_counts().head()

(-33.78815, 151.26654)         1758
(-33.88417103, 151.20235)       685
(-33.88032891, 151.2056904)     407
(-42.84338, 147.29569)          378
(-33.89293336, 151.2022962)     241
Name: Location, dtype: int64

In [75]:
# from 5pm-midnight
df[(df.Datetime.dt.hour > 16)]

# select top locations for this time period
df[(df.Datetime.dt.hour > 16)]['Location'].value_counts().head()

(-33.78815, 151.26654)         1800
(-33.89293336, 151.2022962)     450
(-42.84338, 147.29569)          338
(-33.88417103, 151.20235)       271
(-42.85984, 147.29215)          262
Name: Location, dtype: int64

### Plot on the map!

In [50]:
# access each location by index, and then index inside the tuple
display(df.Location.iloc[5][0])
display(df.Location.iloc[5][1])

-33.884171029999997

151.20235

In [53]:
# plotting code here is just an example - I'm not plotting every combination above
# in this case, we're just plotting all 70 unique locations

map_location = folium.Map(location=[-35.136188, 150.612841], zoom_start = 5)

for x in df.Location.unique():
    folium.Marker((x[0], x[1]), popup='It worked!').add_to(map_location)
map_location

### add these to a heatmap showing frequency at locations

In [66]:
from folium.plugins import HeatMap, HeatMapWithTime

# initialize heatmap, add center point
heatmap_location = folium.Map(location=[-35.136188, 150.612841], zoom_start = 5)

# add the location column to map
HeatMap(df.Location).add_to(heatmap_location)

In [67]:
heatmap_location

### Most Frequent Contacts

In [68]:
df['Comm Identifier'].dropna().value_counts().head(10)

bc0b018604    219
12e3d1b0c9    146
91aba4a113    144
a24a4646d0    133
6bbc17070a     83
a804558e42     62
cd3b394668     56
c22670da93     44
70e1f163d8     39
c521537546     31
Name: Comm Identifier, dtype: int64